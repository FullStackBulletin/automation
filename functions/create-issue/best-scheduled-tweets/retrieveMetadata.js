import { mapLimit } from 'async'
import op from 'object-path'
import debug from 'debug'

const d = debug('retrieveMetadata')

const retrieveMetadataFromLink = metaExtractor => (link, cb) => {
  metaExtractor({ uri: link.id, timeout: 10000 }, (err, metadata) => {
    if (err) {
      // if it's not possible to retrieve metadata ignore the link
      return cb(null)
    }

    const image = op.coalesce(metadata, ['ogImage', 'twitterImageSrc'], null)
    const title = op.get(link, 'og_object.title') || op.coalesce(metadata, ['ogTitle', 'title'])
    const description = op.get(link, 'og_object.description') ||
      op.coalesce(metadata, ['ogDescription', 'twitterDescription', 'description'])

    return cb(null, {
      ...link, image, title, description, metadata
    })
  })
}

export const retrieveMetadata = metaExtractor => (links) => {
  d('Input', links)

  const result = new Promise((resolve, reject) => {
    const limit = 10
    mapLimit(
      links,
      limit,
      retrieveMetadataFromLink(metaExtractor),
      (err, linksWithMetadata) => {
        if (err) {
          return reject(err)
        }
        return resolve(linksWithMetadata)
      }
    )
  })

  d('Output', result)

  return result
}

export default retrieveMetadata
