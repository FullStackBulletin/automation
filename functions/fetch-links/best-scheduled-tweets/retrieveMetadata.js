import { mapLimit } from 'async'
import op from 'object-path'
import debug from 'debug'

const d = debug('retrieveMetadata')

function retrieveMetadataFromLink (metaExtractor) {
  return async function (link) {
    return new Promise((resolve, reject) => {
      metaExtractor({ uri: link.id, timeout: 10000 }, (err, metadata) => {
        if (err) {
          // if it's not possible to retrieve metadata ignore the link
          return resolve(undefined)
        }

        const image = op.coalesce(metadata, ['ogImage', 'twitterImageSrc'], null)
        const title = op.get(link, 'og_object.title') || op.coalesce(metadata, ['ogTitle', 'title'])
        const description = op.get(link, 'og_object.description') ||
          op.coalesce(metadata, ['ogDescription', 'twitterDescription', 'description'])

        return resolve({
          ...link, image, title, description, metadata
        })
      })
    })
  }
}

export const retrieveMetadata = metaExtractor => (links) => {
  d('Input', links)

  const limit = 10
  const result = mapLimit(
    links,
    limit,
    retrieveMetadataFromLink(metaExtractor)
  )

  d('Output', result)

  return result
}

export default retrieveMetadata
