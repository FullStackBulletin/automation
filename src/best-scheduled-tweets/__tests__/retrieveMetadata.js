import { spy } from 'sinon'
import { retrieveMetadata } from '../retrieveMetadata'

test('it should retrieve metadata for a set of given links', async (endTest) => {
  const metadataMap = {
    uri1: {
      ogImage: 'image1'
    },
    uri2: {
      twitterImageSrc: 'image2',
      ogTitle: 'title2',
      ogDescription: 'description2'
    },
    uri3: {
      ogImage: 'image3',
      title: 'title3',
      twitterDescription: 'description3'
    },
    uri4: {
      ogImage: 'image4',
      ogTitle: 'title4',
      description: 'description4'
    }
  }

  const metaExtractor = spy(
    (obj, cb) => setImmediate(() => {
      cb(null, { id: obj.id, ...metadataMap[obj.uri] })
    })
  )

  const links = [
    {
      id: 'uri1',
      og_object: {
        title: 'title1',
        description: 'description1'
      }
    },
    {
      id: 'uri2'
    },
    {
      id: 'uri3'
    },
    {
      id: 'uri4'
    }
  ]

  const data = await (retrieveMetadata(metaExtractor)(links))

  expect(metaExtractor.callCount).toBe(links.length)
  expect(data).toMatchSnapshot()

  endTest()
})

test('it should return undefined if one link fails', async (endTest) => {
  // on someUri2 it will fail
  const metaExtractor = (obj, cb) => setImmediate(
    () => {
      if (obj.uri === 'someUri2') {
        return cb(new Error('some error'))
      }

      return cb(null, obj)
    }
  )

  const links = [1, 2, 3].map(i => ({ id: `someUri${i}` }))
  const result = await (retrieveMetadata(metaExtractor)(links))
  expect(result[1]).toBe(undefined)
  expect(result).toMatchSnapshot()

  endTest()
})

test('it should handle multiple failures', async (endTest) => {
  // on someUri2 it will fail
  const metaExtractor = (obj, cb) => setImmediate(
    () => {
      if (obj.uri === 'someUri2' || obj.uri === 'someUri3') {
        return cb(new Error('some error'))
      }

      return cb(null, obj)
    }
  )

  const links = [1, 2, 3].map(i => ({ id: `someUri${i}` }))
  const result = await (retrieveMetadata(metaExtractor)(links))
  expect(result[1]).toBe(undefined)
  expect(result[2]).toBe(undefined)
  expect(result).toMatchSnapshot()

  endTest()
})
