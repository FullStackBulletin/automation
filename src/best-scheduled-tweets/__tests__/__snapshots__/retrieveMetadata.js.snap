// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should handle multiple failures 1`] = `
Array [
  Object {
    "description": undefined,
    "id": "someUri1",
    "image": null,
    "metadata": Object {
      "timeout": 10000,
      "uri": "someUri1",
    },
    "title": undefined,
  },
  undefined,
  undefined,
]
`;

exports[`it should retrieve metadata for a set of given links 1`] = `
Array [
  Object {
    "description": "description1",
    "id": "uri1",
    "image": "image1",
    "metadata": Object {
      "id": undefined,
      "ogImage": "image1",
    },
    "og_object": Object {
      "description": "description1",
      "title": "title1",
    },
    "title": "title1",
  },
  Object {
    "description": "description2",
    "id": "uri2",
    "image": "image2",
    "metadata": Object {
      "id": undefined,
      "ogDescription": "description2",
      "ogTitle": "title2",
      "twitterImageSrc": "image2",
    },
    "title": "title2",
  },
  Object {
    "description": "description3",
    "id": "uri3",
    "image": "image3",
    "metadata": Object {
      "id": undefined,
      "ogImage": "image3",
      "title": "title3",
      "twitterDescription": "description3",
    },
    "title": "title3",
  },
  Object {
    "description": "description4",
    "id": "uri4",
    "image": "image4",
    "metadata": Object {
      "description": "description4",
      "id": undefined,
      "ogImage": "image4",
      "ogTitle": "title4",
    },
    "title": "title4",
  },
]
`;

exports[`it should return undefined if one link fails 1`] = `
Array [
  Object {
    "description": undefined,
    "id": "someUri1",
    "image": null,
    "metadata": Object {
      "timeout": 10000,
      "uri": "someUri1",
    },
    "title": undefined,
  },
  undefined,
  Object {
    "description": undefined,
    "id": "someUri3",
    "image": null,
    "metadata": Object {
      "timeout": 10000,
      "uri": "someUri3",
    },
    "title": undefined,
  },
]
`;
